src/utils/webrtc/simplewebrtc/peer.js

```diff
@@ -372,22 +400,126 @@ Peer.prototype.end = function() {
 	this.pc.close()
 	this.handleStreamRemoved()
 	this.parent.off('localTrackReplaced', this.handleLocalTrackReplacedBound)
+	this.parent.off('localTrackEnabledChanged', this.handleLocalTrackEnabledChangedBound)
 }
 
 Peer.prototype.handleLocalTrackReplaced = function(newTrack, oldTrack, stream) {
+	this._pendingReplaceTracksQueue.push({ newTrack, oldTrack, stream })
+
+	this._processPendingReplaceTracks()
+}
+
+/**
+ * Process pending replace track actions.
+ *
+ * All the pending replace track actions are executed from the oldest to the
+ * newest, waiting until the previous action was executed before executing the
+ * next one.
+ *
+ * The process may be stopped if the connection is lost, or if a track needs to
+ * be added rather than replaced, which requires a renegotiation. In both cases
+ * the process will start again once the connection is restablished.
+ *
+ * @returns {Promise} a Promise fulfilled when the processing ends; if it was
+ *          completed the resolved value is true, and if it was stopped before
+ *          finishing the resolved value is false.
+ */
+Peer.prototype._processPendingReplaceTracks = function() {
+	if (this._processPendingReplaceTracksPromise) {
+		return this._processPendingReplaceTracksPromise
+	}
+
+	this._processPendingReplaceTracksPromise = this._processPendingReplaceTracksAsync()
+
+	// For compatibility with older browsers "finally" should not be used on
+	// Promises.
+	this._processPendingReplaceTracksPromise.then(() => {
+		this._processPendingReplaceTracksPromise = null
+	}).catch(() => {
+		this._processPendingReplaceTracksPromise = null
+	})
+
+	return this._processPendingReplaceTracksPromise
+}
+
+Peer.prototype._processPendingReplaceTracksAsync = async function() {
+	for (let i = 0; i < this._pendingReplaceTracksQueue.length; i++) {
+		if (this.pc.iceConnectionState === 'new') {
+			// Do not replace the tracks when the connection has not started
+			// yet, as Firefox can get "stuck" and not replace the tracks
+			// even if tried later again once connected.
+			return false
+		}
+
+		const pending = this._pendingReplaceTracksQueue[i]
+
+		try {
+			await this._replaceTrack(pending.newTrack, pending.oldTrack, pending.stream)
+		} catch (exception) {
+			// If the track is added instead of replaced a renegotiation
+			// will be needed, so stop replacing tracks.
+			return false
+		} finally {
```

Comment:
didn't you just say above that we should not use finally ? :wink: 

now with IE being phased out I guess it's ok to use it ?

Reply:
Yes, but with Promises; `try/catch/finally` is safe to use ;-)

Regarding [browser compatibility](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#browser_compatibility) we still list [Firefox 52 and Chromium 49 as supported](https://github.com/nextcloud/spreed/blob/cc168f43a965504c9269dec3f840ad41f219e151/README.md#supported-browsers) :shrug: